source/lac/cuda_precondition.cu

```diff
@@ -0,0 +1,1844 @@
+    , policy_M(CUSPARSE_SOLVE_POLICY_USE_LEVEL)
+    , n_rows(0)
+    , n_nonzero_elements(0)
+  {
+    cusparseStatus_t status;
+    // step 1: create a descriptor which contains
+    // - matrix M is base-0
+    // - matrix L is base-0
+    // - matrix L is lower triangular
+    // - matrix L has non-unit diagonal
+    status = cusparseCreateMatDescr(&descr_M);
+    AssertCusparse(status);
+    status = cusparseSetMatIndexBase(descr_M, CUSPARSE_INDEX_BASE_ZERO);
+    AssertCusparse(status);
+    status = cusparseSetMatType(descr_M, CUSPARSE_MATRIX_TYPE_GENERAL);
+    AssertCusparse(status);
+
+    status = cusparseCreateMatDescr(&descr_L);
+    AssertCusparse(status);
+    status = cusparseSetMatIndexBase(descr_L, CUSPARSE_INDEX_BASE_ZERO);
+    AssertCusparse(status);
+    status = cusparseSetMatType(descr_L, CUSPARSE_MATRIX_TYPE_GENERAL);
+    AssertCusparse(status);
+    status = cusparseSetMatFillMode(descr_L, CUSPARSE_FILL_MODE_LOWER);
+    AssertCusparse(status);
+    status = cusparseSetMatDiagType(descr_L, CUSPARSE_DIAG_TYPE_NON_UNIT);
+    AssertCusparse(status);
+
+    // step 2: create a empty info structure
+    // we need one info for csric02 and two info's for csrsv2
+    status = cusparseCreateCsric02Info(&info_M);
+    AssertCusparse(status);
+    status = cusparseCreateCsrsv2Info(&info_L);
+    AssertCusparse(status);
+    status = cusparseCreateCsrsv2Info(&info_Lt);
+    AssertCusparse(status);
+  }
+
+
+
+  template <typename Number>
+  PreconditionIC<Number>::~PreconditionIC()
+  {
+    // step 8: free resources
+    cusparseStatus_t status = cusparseDestroyMatDescr(descr_M);
+    AssertNothrowCusparse(status);
+
+    status = cusparseDestroyMatDescr(descr_L);
+    AssertNothrowCusparse(status);
+
+    status = cusparseDestroyCsric02Info(info_M);
+    AssertNothrowCusparse(status);
+
+    status = cusparseDestroyCsrsv2Info(info_L);
+    AssertNothrowCusparse(status);
+
+    status = cusparseDestroyCsrsv2Info(info_Lt);
+    AssertNothrowCusparse(status);
+  }
+
+
+
+  template <typename Number>
+  void
+  PreconditionIC<Number>::initialize(const SparseMatrix<Number> &A,
+                                     const AdditionalData &additional_data)
+  {
+    if (additional_data.use_level_analysis)
+      {
+        policy_L  = CUSPARSE_SOLVE_POLICY_USE_LEVEL;
+        policy_Lt = CUSPARSE_SOLVE_POLICY_USE_LEVEL;
+        policy_M  = CUSPARSE_SOLVE_POLICY_USE_LEVEL;
+      }
+    else
+      {
+        policy_L  = CUSPARSE_SOLVE_POLICY_NO_LEVEL;
+        policy_Lt = CUSPARSE_SOLVE_POLICY_NO_LEVEL;
+        policy_M  = CUSPARSE_SOLVE_POLICY_NO_LEVEL;
+      }
+
+    n_rows             = A.m();
+    n_nonzero_elements = A.n_nonzero_elements();
+    AssertDimension(A.m(), A.n());
+
+    matrix_pointer                      = &A;
+    const auto          cusparse_matrix = A.get_cusparse_matrix();
+    const Number *const A_val_dev       = std::get<0>(cusparse_matrix);
+
+    // create a copy of the matrix entries
```

Comment:
Why do you do this? Is the data overwritten?

Reply:
Yes, it is.